File: core/protocol.py
from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
from typing import List, Sequence

from .models import AncMode, CodecMode, EqMode, FirmwareVersion


MAGIC = 0x4E


class Op(IntEnum):
    VERSION = 0x0003
    LEGACY_CODEC_LHDC = 0x0004
    BATTERY = 0x0005
    ANC_SET = 0x0201
    ANC_QUERY = 0x0101
    EQ_SET = 0x0207
    EQ_QUERY = 0x0107
    GAME_MODE_SET = 0x0208
    GAME_MODE_QUERY = 0x0108
    LOW_LATENCY_SET = 0x0206
    LOW_LATENCY_QUERY = 0x0106
    DUAL_CONN_SET = 0x0205
    DUAL_CONN_QUERY = 0x0105
    IN_EAR_SET = 0x0209
    IN_EAR_QUERY = 0x0109
    CODEC = 0x0204
    WIND_SUPPRESSION_SET = 0x02E1
    WIND_SUPPRESSION_QUERY = 0x01E1
    FULL_STATE = 0x0103


@dataclass(slots=True)
class ParsedMessage:
    op_code: int
    payload: bytes


@dataclass(slots=True)
class ParsedStateUpdate:
    anc_mode: AncMode | None = None
    eq_mode: EqMode | None = None
    left_battery: int | None = None
    right_battery: int | None = None
    case_battery: int | None = None
    game_mode_enabled: bool | None = None
    low_latency_enabled: bool | None = None
    firmware: FirmwareVersion | None = None


class PacketStreamParser:
    def __init__(self) -> None:
        self._buffer = bytearray()

    def feed(self, data: bytes) -> List[ParsedMessage]:
        self._buffer.extend(data)
        messages: List[ParsedMessage] = []

        while True:
            start = self._find_magic()
            if start < 0:
                self._buffer.clear()
                break
            if start > 0:
                del self._buffer[:start]
            if len(self._buffer) < 6:
                break

            payload_length = self._buffer[1] | (self._buffer[2] << 8)
            packet_length = payload_length + 3
            if len(self._buffer) < packet_length:
                break

            packet = bytes(self._buffer[:packet_length])
            del self._buffer[:packet_length]

            op_code = packet[4] | (packet[5] << 8)
            messages.append(ParsedMessage(op_code=op_code, payload=packet[6:]))

        return messages

    def _find_magic(self) -> int:
        for index, value in enumerate(self._buffer):
            if value == MAGIC:
                return index
        return -1


def build_command(op_code: int, *params: int) -> bytes:
    payload_length = 3 + len(params)
    packet = bytearray(3 + payload_length)
    packet[0] = MAGIC
    packet[1] = payload_length & 0xFF
    packet[2] = (payload_length >> 8) & 0xFF
    packet[3] = 0x00
    packet[4] = op_code & 0xFF
    packet[5] = (op_code >> 8) & 0xFF
    for index, param in enumerate(params, start=6):
        packet[index] = param & 0xFF
    return bytes(packet)


def query_firmware() -> bytes:
    return build_command(Op.VERSION)


def query_battery() -> bytes:
    return build_command(Op.BATTERY)


def query_anc() -> bytes:
    return build_command(Op.ANC_QUERY)


def set_anc(mode: AncMode) -> bytes:
    return build_command(Op.ANC_SET, mode.value, 0x00)


def query_eq() -> bytes:
    return build_command(Op.EQ_QUERY)


def set_eq(mode: EqMode) -> bytes:
    return build_command(Op.EQ_SET, mode.value)


def set_codec(mode: CodecMode) -> bytes:
    return build_command(Op.CODEC, mode.value)


def set_legacy_codec_lhdc(enabled: bool) -> bytes:
    return build_command(Op.LEGACY_CODEC_LHDC, 0x01 if enabled else 0x00)


def query_game_mode() -> bytes:
    return build_command(Op.GAME_MODE_QUERY)


def set_game_mode(enabled: bool) -> bytes:
    return build_command(Op.GAME_MODE_SET, 0x01 if enabled else 0x00)


def query_low_latency() -> bytes:
    return build_command(Op.LOW_LATENCY_QUERY)


def set_low_latency(enabled: bool) -> bytes:
    return build_command(Op.LOW_LATENCY_SET, 0x01 if enabled else 0x00)


def parse_state_update(message: ParsedMessage) -> ParsedStateUpdate:
    payload = message.payload
    if message.op_code == Op.BATTERY and len(payload) >= 3:
        return ParsedStateUpdate(
            left_battery=payload[0],
            right_battery=payload[1],
            case_battery=(payload[2] if payload[2] != 0 else None),
        )
    if message.op_code == Op.ANC_QUERY and len(payload) >= 1:
        return ParsedStateUpdate(anc_mode=AncMode.from_value(payload[0]))
    if message.op_code == Op.EQ_QUERY and len(payload) >= 1:
        return ParsedStateUpdate(eq_mode=EqMode.from_value(payload[0]))
    if message.op_code == Op.GAME_MODE_QUERY and len(payload) >= 1:
        return ParsedStateUpdate(game_mode_enabled=(payload[0] == 0x01))
    if message.op_code == Op.LOW_LATENCY_QUERY and len(payload) >= 1:
        return ParsedStateUpdate(low_latency_enabled=(payload[0] == 0x01))
    if message.op_code == Op.VERSION and len(payload) >= 2:
        return ParsedStateUpdate(firmware=FirmwareVersion(main=payload[1], sub=payload[0]))
    return ParsedStateUpdate()


def paced_startup_queries() -> Sequence[bytes]:
    return (
        query_firmware(),
        query_battery(),
        query_anc(),
        query_eq(),
        query_game_mode(),
        query_low_latency(),
    )

File: src/protocol.rs
use crate::models::{AncMode, CodecMode, EqMode, FirmwareVersion};

pub const MAGIC: u8 = 0x4e;
const HEADER_LENGTH: usize = 6;
const MIN_PAYLOAD_LENGTH: usize = 3;
const MAX_PAYLOAD_LENGTH: usize = 4096;

#[repr(u16)]
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Op {
    Version = 0x0003,
    LegacyCodecLhdc = 0x0004,
    Battery = 0x0005,
    AncSet = 0x0201,
    AncQuery = 0x0101,
    EqSet = 0x0207,
    EqQuery = 0x0107,
    GameModeSet = 0x0208,
    GameModeQuery = 0x0108,
    LowLatencySet = 0x0206,
    LowLatencyQuery = 0x0106,
    DualConnSet = 0x0205,
    DualConnQuery = 0x0105,
    InEarSet = 0x0209,
    InEarQuery = 0x0109,
    Codec = 0x0204,
    WindSuppressionSet = 0x02e1,
    WindSuppressionQuery = 0x01e1,
    FullState = 0x0103,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ParsedMessage {
    pub op_code: u16,
    pub payload: Vec<u8>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct ParsedStateUpdate {
    pub anc_mode: Option<AncMode>,
    pub eq_mode: Option<EqMode>,
    pub left_battery: Option<u8>,
    pub right_battery: Option<u8>,
    pub case_battery: Option<Option<u8>>,
    pub game_mode_enabled: Option<bool>,
    pub low_latency_enabled: Option<bool>,
    pub firmware: Option<FirmwareVersion>,
}

#[derive(Debug, Default)]
pub struct PacketStreamParser {
    buffer: Vec<u8>,
}

impl PacketStreamParser {
    pub fn feed(&mut self, data: &[u8]) -> Vec<ParsedMessage> {
        self.buffer.extend_from_slice(data);
        let mut messages = Vec::new();

        loop {
            let Some(start) = self.buffer.iter().position(|byte| *byte == MAGIC) else {
                self.buffer.clear();
                break;
            };
            if start > 0 {
                self.buffer.drain(..start);
            }
            if self.buffer.len() < HEADER_LENGTH {
                break;
            }

            let payload_length = u16::from_le_bytes([self.buffer[1], self.buffer[2]]) as usize;
            if !(MIN_PAYLOAD_LENGTH..=MAX_PAYLOAD_LENGTH).contains(&payload_length) {
                self.buffer.remove(0);
                continue;
            }

            let packet_length = payload_length + 3;
            if self.buffer.len() < packet_length {
                break;
            }

            let op_code = u16::from_le_bytes([self.buffer[4], self.buffer[5]]);
            let payload = self.buffer[HEADER_LENGTH..packet_length].to_vec();
            self.buffer.drain(..packet_length);
            messages.push(ParsedMessage { op_code, payload });
        }

        messages
    }
}

pub fn build_command(op_code: u16, params: &[u8]) -> Vec<u8> {
    let payload_length = MIN_PAYLOAD_LENGTH + params.len();
    let mut packet = Vec::with_capacity(payload_length + 3);
    packet.push(MAGIC);
    packet.extend_from_slice(&(payload_length as u16).to_le_bytes());
    packet.push(0);
    packet.extend_from_slice(&op_code.to_le_bytes());
    packet.extend_from_slice(params);
    packet
}

pub fn query_firmware() -> Vec<u8> {
    build_command(Op::Version as u16, &[])
}

pub fn query_battery() -> Vec<u8> {
    build_command(Op::Battery as u16, &[])
}

pub fn query_anc() -> Vec<u8> {
    build_command(Op::AncQuery as u16, &[])
}

pub fn set_anc(mode: AncMode) -> Vec<u8> {
    build_command(Op::AncSet as u16, &[mode.value(), 0])
}

pub fn query_eq() -> Vec<u8> {
    build_command(Op::EqQuery as u16, &[])
}

pub fn set_eq(mode: EqMode) -> Vec<u8> {
    build_command(Op::EqSet as u16, &[mode.value()])
}

pub fn set_codec(mode: CodecMode) -> Vec<u8> {
    build_command(Op::Codec as u16, &[mode.value()])
}

pub fn set_legacy_codec_lhdc(enabled: bool) -> Vec<u8> {
    build_command(Op::LegacyCodecLhdc as u16, &[u8::from(enabled)])
}

pub fn query_game_mode() -> Vec<u8> {
    build_command(Op::GameModeQuery as u16, &[])
}

pub fn set_game_mode(enabled: bool) -> Vec<u8> {
    build_command(Op::GameModeSet as u16, &[u8::from(enabled)])
}

pub fn query_low_latency() -> Vec<u8> {
    build_command(Op::LowLatencyQuery as u16, &[])
}

pub fn set_low_latency(enabled: bool) -> Vec<u8> {
    build_command(Op::LowLatencySet as u16, &[u8::from(enabled)])
}

pub fn parse_state_update(message: &ParsedMessage) -> ParsedStateUpdate {
    let payload = &message.payload;
    match message.op_code {
        op if op == Op::Battery as u16 && payload.len() >= 3 => ParsedStateUpdate {
            left_battery: Some(payload[0]),
            right_battery: Some(payload[1]),
            case_battery: Some((payload[2] != 0).then_some(payload[2])),
            ..Default::default()
        },
        op if op == Op::AncQuery as u16 && !payload.is_empty() => ParsedStateUpdate {
            anc_mode: Some(payload[0].into()),
            ..Default::default()
        },
        op if op == Op::EqQuery as u16 && !payload.is_empty() => ParsedStateUpdate {
            eq_mode: Some(payload[0].into()),
            ..Default::default()
        },
        op if op == Op::GameModeQuery as u16 && !payload.is_empty() => ParsedStateUpdate {
            game_mode_enabled: Some(payload[0] == 1),
            ..Default::default()
        },
        op if op == Op::LowLatencyQuery as u16 && !payload.is_empty() => ParsedStateUpdate {
            low_latency_enabled: Some(payload[0] == 1),
            ..Default::default()
        },
        op if op == Op::Version as u16 && payload.len() >= 2 => ParsedStateUpdate {
            firmware: Some(FirmwareVersion::new(payload[1], payload[0])),
            ..Default::default()
        },
        _ => ParsedStateUpdate::default(),
    }
}

pub fn paced_startup_queries() -> Vec<Vec<u8>> {
    vec![
        query_firmware(),
        query_battery(),
        query_anc(),
        query_eq(),
        query_game_mode(),
        query_low_latency(),
    ]
}
